lib/ansible/modules/database/postgresql/postgresql_set.py

```diff
@@ -0,0 +1,381 @@
+    default: prefer
+  ssl_rootcert:
+    description:
+    - Specifies the name of a file containing SSL certificate authority (CA)
+      certificate(s).
+    - If the file exists, the server's certificate will be
+      verified to be signed by one of these authorities.
+    type: str
+notes:
+- Check_mode is not supported because ALTER SYSTEM can't be run inside a transaction block
+  U(https://www.postgresql.org/docs/current/sql-altersystem.html).
+- Supported version of PostgreSQL is 9.4 and later.
+- For some parameters restart of PostgreSQL server is required.
+  See official documentation U(https://www.postgresql.org/).
+- The default authentication assumes that you are either logging in as or
+  sudo'ing to the postgres account on the host.
+- This module uses psycopg2, a Python PostgreSQL database adapter. You must
+  ensure that psycopg2 is installed on the host before using this module. If
+  the remote host is the PostgreSQL server (which is the default case), then
+  PostgreSQL must also be installed on the remote host. For Ubuntu-based
+  systems, install the postgresql, libpq-dev, and python-psycopg2 packages
+  on the remote host before using this module.
+requirements: [ psycopg2 ]
+author:
+- Andrew Klychkov (@Andersson007)
+'''
+
+EXAMPLES = r'''
+- name: Restore wal_keep_segments parameter to initial state
+  postgresql_set:
+    name: wal_keep_segments
+    reset: yes
+
+# Set work_mem parameter to 32MB and show what's been changed and restart is required or not
+# (output example: "msg": "work_mem 4MB >> 64MB restart_req: False")
+- postgresql_set:
+    name: work_mem
+    value: 32mb
+    register: set_value
+
+- debug:
+    msg: "{{ set_value.name }} {{ set_value.prev_val }} >> {{ set_value.cur_value }} restart_req: {{ set_value.restart_required }}"
+  when: set_value.changed
+# Ensure that the restart of PostgreSQL serever must be required for some parameters.
+# In this situation you see the same parameter in prev_val and cur_value, but 'changed=True'
+# (Of course, if you passed the value that was different from the current server setting).
+
+- name: Set log_min_duration_statement parameter to 1 second
+  postgresql_set:
+    name: log_min_duration_statement
+    value: 1s
+
+- name: Set wal_log_hints parameter to default value (remove parameter from postgresql.auto.conf)
+  postgresql_set:
+    name: wal_log_hints
+    value: default
+'''
+
+RETURN = r'''
+name:
+  description: Name of PostgreSQL server parameter.
+  returned: always
+  type: str
+  sample: 'shared_buffers'
+restart_required:
+  description: Information about parameter current state.
+  returned: always
+  type: bool
+  sample: true
+prev_val:
+  description: Information about previous state of the parameter.
+  returned: always
+  type: str
+  sample: '4MB'
+cur_val:
+  description: Information about current state of the parameter.
+  returned: always
+  type: str
+  sample: '64MB'
```

Comment:
Values we return are in bytes.

Reply:
@dagwieers I'm afraid the implementation of this suggestion makes the code very complex because a parameter can have a value like in following output from PostgreSQL system view pg_settings:
```
                name                 |                setting                 | unit 
-------------------------------------+----------------------------------------+------
 allow_system_table_mods             | off                                    | 
 application_name                    | psql                                   | 
 archive_command                     | (disabled)                             | 
 archive_mode                        | off                                    | 
 archive_timeout                     | 0                                      | s
 array_nulls                         | on                                     | 
 authentication_timeout              | 60                                     | s
 autovacuum                          | on                                     | 
 autovacuum_analyze_scale_factor     | 0.1                                    | 
 autovacuum_analyze_threshold        | 50                                     | 
 autovacuum_freeze_max_age           | 200000000                              | 
 autovacuum_max_workers              | 3                                      | 
 autovacuum_multixact_freeze_max_age | 400000000                              | 
 autovacuum_naptime                  | 60                                     | s
 autovacuum_vacuum_cost_delay        | 20                                     | ms
 autovacuum_vacuum_cost_limit        | -1                                     | 
 autovacuum_vacuum_scale_factor      | 0.2                                    | 
 autovacuum_vacuum_threshold         | 50                                     | 
 autovacuum_work_mem                 | -1                                     | kB
 backend_flush_after                 | 0                                      | 8kB
 backslash_quote                     | safe_encoding                          | 
 bgwriter_delay                      | 200                                    | ms
 bgwriter_flush_after                | 64                                     | 8kB
 bgwriter_lru_maxpages               | 100                                    | 
 bgwriter_lru_multiplier             | 2                                      | 
 block_size                          | 8192                                   | 
 bonjour                             | off                                    | 
 bonjour_name                        |                                        | 
 bytea_output                        | hex                                    | 
 check_function_bodies               | on                                     | 
 checkpoint_completion_target        | 0.5                                    | 
 checkpoint_flush_after              | 32                                     | 8kB
 checkpoint_timeout                  | 300                                    | s
 checkpoint_warning                  | 30                                     | s
 client_encoding                     | UTF8                                   | 
 client_min_messages                 | notice                                 | 
 cluster_name                        |                                        | 
 commit_delay                        | 0                                      | 
 commit_siblings                     | 5                                      | 
 config_file                         | /var/lib/pgsql/10/data/postgresql.conf | 
```
And this is a small part... there are other cases like 'logical', 'hot standby', etc

Reply:
We (dba) usually run "SHOW param_name" to know about the current setting value. I use this in the module output to present it.

Reply:
I understand, but one of the reasons for returning information, is to be able to use it in future tasks. If you return 64MB, it is impossible to manipulate that data (say you want only half).

That's why it is better to return absolute values (in bytes) so you can decide whether values ought to be MB or MiB, and have full control over what you sent back.

Reply:
I know that it's better. Ok, I should think of how to implement this.

Reply:
My main problem is that whatever we ship now, is set in stone. We cannot improve this in the future easily (except if we return 2 variables, one for backward compatibility). So interfaces are very important once people depend on it.

Reply:
Is it OK if I add the next returned value?
```
170 setting:
171   description:
172   - Dictionary that contains the current parameter value (at the time of playbook finish).
173   - Pay attention that for real change some parameters restart of PostgreSQL server is required.
174   returned: always
175   type: dict
176   sample: { "value": '67108864', "unit": "b" }
```

Reply:
Solved. I add this:
```
setting:
  description:
  - Dictionary that contains the current parameter value (at the time of playbook finish).
  - Pay attention that for real change some parameters restart of PostgreSQL server is required.
  returned: always
  type: dict
  sample: { "value": 67108864, "unit": "b" }
```
OK?
(every time when I change my modules I find something new useful that can be added or code that can be removed ;)